//  @file   config.proto
//  @brief  configuration for factorization machine
package dmlc.fm;

message Config {
  //
  // data
  //
  optional string train_data = 1;
  optional string val_data = 2;
  optional string test_data = 3;
  optional string data_format = 4 [default = "libsvm"];

  // partition a file into npart
  optional int32 num_parts_per_file = 5 [default = 1];

  //
  // model
  //
  optional string model_in = 10;
  optional string model_out = 11;

  optional float lambda_l2 = 20;
  optional float lambda_l1 = 21;

  message Embedding {
    optional int32 dim = 1;
    optional int32 threshold = 2;
    optional float lambda_l2 = 3;
    optional float lr_eta = 4 [default = .01];
    optional float lr_beta = 5 [default = 1];
    optional float init_max = 6 [default = .01];
    optional float init_min = 7 [default = -.01];
  }

  repeated Embedding embedding = 22;

  //
  // optimization methods
  //

  // the size of minibatch for online methods
  optional int32 minibatch = 31 [default = 1000];

  // the maximal number of data passes
  optional int32 max_data_pass = 32 [default = 10];

  // print the progress every n sec
  optional float disp_itv = 33 [default = 1];

  // learning rate, often in the format lr_eta / (lr_beta + x), where x is
  // user-defined, such as sqrt(iter), or the cumulative gradient on adagrad
  optional float lr_eta = 41;
  optional float lr_beta = 42;

  //
  // to reduce sync cost
  //

  // maximal allowed delay during synchronization
  optional int32 max_delay = 50 [default = 0];
  // communication filters
  optional bool key_cache = 51 [default = true];
  optional bool msg_compression = 52 [default = true];
  // 0 means no fixing
  optional int32 fixed_bytes = 53 [default = 1];

  optional int32 num_threads = 60 [default = 2];
}
